{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "263a6301",
   "metadata": {},
   "outputs": [],
   "source": [
    "# File paths\n",
    "genes_file = \"input-files/tip60-genes.txt\"\n",
    "# Load gene list\n",
    "with open(genes_file, \"r\") as f:\n",
    "    genes = [line.strip() for line in f.readlines()]\n",
    "\n",
    "illumina_deseq = \"DESeq2_results_shrunk.csv\" # deseq2 illumina\n",
    "pacbio_deseq_isoforms = 'DESeq2_results_shrunk_isoforms.csv' # deseq2 pacbio\n",
    "classifications = 'collapsed_classification.filtered_lite_classification.txt' # pigeon classify\n",
    "gff = 'gencode.v39.annotation.sorted.gtf' # this will enable cross referencesing the files\n",
    "p_cutoff = 0.05\n",
    "\n",
    "\n",
    "## Parse GFF to get lookup tables\n",
    "gff_lookup = {} # gene_id to gene name\n",
    "gff_lookup_reverse = {} # gen_name to gene_id\n",
    "transcript_to_gene = {}\n",
    "gene_to_transcripts = {}\n",
    "\n",
    "for line in open(gff):\n",
    "    if line[0] == '#':\n",
    "        continue\n",
    "    elements = line.rstrip().split('\\t')\n",
    "    info = elements[-1].split(';')\n",
    "    \n",
    "    if elements[2] == 'gene':\n",
    "        gene_name = ''\n",
    "        gene_id = ''\n",
    "        for thing in info:\n",
    "            if ' ' not in thing:\n",
    "                continue\n",
    "            if thing.split(' ')[1] == 'gene_name':\n",
    "                gene_name = thing.split(' ')[2].replace('\"','')\n",
    "            if thing.split(' ')[0] == 'gene_id':\n",
    "                gene_id = thing.split(' ')[1].replace('\"','')\n",
    "        gff_lookup[gene_id] = gene_name\n",
    "        gff_lookup_reverse[gene_name] = gene_id\n",
    "    \n",
    "    if elements[2] == 'transcript':\n",
    "        transcript_id = ''\n",
    "        gene_id = ''\n",
    "        for thing in info:\n",
    "            if ' ' not in thing:\n",
    "                continue\n",
    "            if thing.split(' ')[1] == 'transcript_id':\n",
    "                transcript_id = thing.split(' ')[2]\n",
    "            if thing.split(' ')[0] == 'gene_id':\n",
    "                gene_id = thing.split(' ')[1]\n",
    "\n",
    "        if gene_id != '' and transcript_id != '':\n",
    "            transcript_to_gene.setdefault(transcript_id, '')\n",
    "            transcript_to_gene[transcript_id] = gene_id\n",
    "            gene_to_transcripts.setdefault(gene_id, set())  # Initialize as a set\n",
    "            gene_to_transcripts[gene_id].add(transcript_id)\n",
    "\n",
    "\n",
    "# Parse isoseq classifications\n",
    "isoform_to_gene = {}\n",
    "isoform_classification  = {}\n",
    "for line in open(classifications):\n",
    "    elements = line.rstrip().split('\\t')\n",
    "    if elements[0] == 'isoform' or elements[0] == '\"\"':\n",
    "        continue\n",
    "    isoform  = elements[0].replace('\"','')\n",
    "    ass_gene = elements[6]\n",
    "    transcript = elements[7]\n",
    "    category = elements[14]\n",
    "    \n",
    "    isoform_classification[isoform] = category\n",
    "    \n",
    "    try:\n",
    "        gene = gff_lookup_reverse[ass_gene]\n",
    "    except KeyError:\n",
    "        gene = 'Unknown'\n",
    "    isoform_to_gene.setdefault(isoform, ['','', ''])\n",
    "    isoform_to_gene[isoform] = [gene, ass_gene, transcript]\n",
    "            \n",
    "            \n",
    "## parse illumina deseq\n",
    "illumina_lookup = {}\n",
    "for line in open(illumina_deseq):\n",
    "    elements = line.rstrip().split(',')\n",
    "    gene = elements[0].replace('\"','')\n",
    "    illumina_lookup[gene] = elements\n",
    "\n",
    "pacbio_lookup = {}\n",
    "for line in open(pacbio_deseq_isoforms):\n",
    "    elements=line.rstrip().split(',')\n",
    "    if elements[1] == 'baseMean' or elements[0] == '\"\"':\n",
    "        continue\n",
    "    isoform = elements[0].replace('\"','')\n",
    "    gene = isoform_to_gene[isoform][0]\n",
    "    \n",
    "    if gene in gff_lookup.keys():\n",
    "        gene_name = gff_lookup[gene]\n",
    "        pacbio_lookup.setdefault(gene_name, [])\n",
    "        pacbio_lookup[gene_name].append(elements)\n",
    "    else:\n",
    "        gene_name = 'Unknown'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "27d2cc30",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene\tlog2FoldChange\tpadj\tnum_isoforms\tsignificant_up\tsignificant_down\tcategories\n",
      "KAT5\t-0.410763045\t0.000225465\t13\t0\t2\treference_match|reference_match|alternative_5end|alternative_5end|alternative_3end|reference_match|combination_of_known_junctions|alternative_3end5end|alternative_5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment\n",
      "EP400\t-0.280616191\t0.038109071\t1\t0\t0\tcombination_of_known_junctions\n",
      "TRRAP\t-0.138325654\t0.414363477\t4\t0\t0\tinternal_fragment|internal_fragment|internal_fragment|internal_fragment\n",
      "BRD8\t0.157954396\t0.35760197\t15\t0\t0\treference_match|combination_of_known_junctions|combination_of_known_junctions|reference_match|combination_of_known_junctions|reference_match|combination_of_known_junctions|combination_of_known_junctions|combination_of_known_junctions|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|reference_match|3prime_fragment\n",
      "EPC1\t0.169092142\t0.139968173\t6\t0\t1\talternative_3end|5prime_fragment|5prime_fragment|5prime_fragment|reference_match|reference_match\n",
      "EPC2\t0.064374159\t0.701234555\t4\t0\t0\treference_match|3prime_fragment|3prime_fragment|3prime_fragment\n",
      "MBTD1\t-0.181013499\t0.07057304\t2\t0\t0\t5prime_fragment|at_least_one_novel_splicesite\n",
      "DMAP1\t-0.145453953\t0.279852815\t13\t0\t2\treference_match|combination_of_known_junctions|at_least_one_novel_splicesite|reference_match|combination_of_known_splicesites|combination_of_known_junctions|intron_retention|intron_retention|at_least_one_novel_splicesite|combination_of_known_junctions|alternative_5end|3prime_fragment|alternative_5end\n",
      "ING3\t0.024888729\t0.923187566\t8\t0\t2\talternative_3end|alternative_3end|reference_match|reference_match|alternative_3end|alternative_3end|alternative_3end|alternative_3end\n",
      "VPS72\t-0.10319127\t0.650861628\t13\t0\t3\talternative_3end|alternative_3end|at_least_one_novel_splicesite|combination_of_known_splicesites|at_least_one_novel_splicesite|at_least_one_novel_splicesite|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|combination_of_known_splicesites|alternative_3end5end|alternative_3end\n",
      "RUVBL2\t0.111006825\t0.368152642\t15\t0\t0\tcombination_of_known_junctions|alternative_5end|combination_of_known_junctions|alternative_5end|reference_match|combination_of_known_junctions|at_least_one_novel_splicesite|at_least_one_novel_splicesite|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment\n",
      "RUVBL1\t0.401894035\t0.000283264\t14\t0\t0\treference_match|alternative_3end|alternative_3end5end|at_least_one_novel_splicesite|combination_of_known_splicesites|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|combination_of_known_splicesites|3prime_fragment\n",
      "ACTL6A\t0.131775332\t0.238432469\t13\t0\t0\treference_match|alternative_3end|reference_match|alternative_3end|alternative_3end|alternative_5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|alternative_3end5end\n",
      "ACTB\t0.158245079\t0.061976246\t33\t0\t1\treference_match|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|alternative_5end|alternative_3end|intron_retention|alternative_3end|alternative_5end|intron_retention|alternative_5end|alternative_5end|alternative_5end|reference_match|intron_retention|alternative_5end|alternative_5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|intron_retention|3prime_fragment|3prime_fragment|at_least_one_novel_splicesite|3prime_fragment|3prime_fragment|intron_retention|intron_retention|alternative_3end|alternative_3end|alternative_3end\n",
      "MORF4L1\t0.238586326\t0.082911859\t33\t0\t0\talternative_3end5end|alternative_3end5end|alternative_3end|alternative_3end|alternative_3end5end|alternative_3end5end|combination_of_known_junctions|5prime_fragment|alternative_3end|alternative_3end5end|alternative_3end5end|alternative_3end5end|alternative_3end5end|at_least_one_novel_splicesite|at_least_one_novel_splicesite|combination_of_known_junctions|combination_of_known_splicesites|combination_of_known_splicesites|combination_of_known_junctions|alternative_3end|alternative_3end5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment\n",
      "MORF4L2\t-0.59104297\t1.92E-11\t18\t0\t6\talternative_3end5end|combination_of_known_splicesites|alternative_5end|alternative_3end|alternative_3end5end|alternative_3end|reference_match|alternative_3end|combination_of_known_junctions|alternative_5end|combination_of_known_splicesites|alternative_3end|alternative_5end|alternative_3end5end|reference_match|3prime_fragment|alternative_3end5end|3prime_fragment\n",
      "YEATS4\t-0.044416731\t0.783745789\t7\t0\t3\talternative_3end|alternative_3end|alternative_3end|at_least_one_novel_splicesite|alternative_3end5end|3prime_fragment|3prime_fragment\n",
      "MRGBP\t0.172190429\t0.417494383\t3\t0\t0\talternative_3end|alternative_3end5end|3prime_fragment\n",
      "MEAF6\t0.113594461\t0.368049763\t31\t0\t9\talternative_3end|alternative_3end|reference_match|alternative_3end|alternative_3end|alternative_3end|alternative_3end|alternative_3end|alternative_3end|alternative_3end5end|alternative_3end5end|alternative_3end5end|3prime_fragment|3prime_fragment|3prime_fragment|alternative_3end|alternative_3end5end|alternative_3end|alternative_3end|reference_match|alternative_3end5end|alternative_3end5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|alternative_3end|alternative_3end|alternative_3end|5prime_fragment\n",
      "H2AZ1\t0.040048488\t0.856458689\t10\t0\t0\talternative_5end|alternative_5end|3prime_fragment|3prime_fragment|intron_retention|intron_retention|alternative_5end|at_least_one_novel_splicesite|alternative_5end|alternative_5end\n",
      "MYCL\t0.378667105\t3.24E-09\t15\t0\t0\treference_match|alternative_5end|alternative_5end|reference_match|alternative_5end|alternative_5end|alternative_5end|alternative_5end|at_least_one_novel_splicesite|alternative_5end|alternative_3end5end|reference_match|alternative_5end|alternative_3end|alternative_3end5end\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print ('Gene\\tlog2FoldChange\\tpadj\\tnum_isoforms\\tsignificant_up\\tsignificant_down\\tcategories')\n",
    "\n",
    "for cur_gene in genes:\n",
    "    \n",
    "    illumina_fold = illumina_lookup[cur_gene][2]\n",
    "    illumina_p = illumina_lookup[cur_gene][5]\n",
    "    \n",
    "    num_isoforms = len(pacbio_lookup[cur_gene])\n",
    "    categories = []\n",
    "    num_up = 0\n",
    "    num_down = 0\n",
    "    \n",
    "    for isoform_results in pacbio_lookup[cur_gene]:\n",
    "\n",
    "        if float(isoform_results[5]) <= 0.05:\n",
    "            if float(elements[2]) > 0:\n",
    "                num_up += 1\n",
    "            else:\n",
    "                num_down += 1\n",
    "        isoform = isoform_results[0].replace('\"','')\n",
    "        classification = isoform_classification[isoform]\n",
    "        categories.append(classification)\n",
    "    print (cur_gene + '\\t' + illumina_fold + '\\t' + illumina_p + '\\t' + str(num_isoforms) + '\\t' + str(num_up) + '\\t' + str(num_down) + '\\t'  +'|'.join(categories))\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "609fb551",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene\tlog2FoldChange\tpadj\tnum_isoforms\tsignificant_up\tsignificant_down\tcategories\n",
      "KAT5\t-0.410763045\t0.000225465\t12\t0\t2\talternative_5end|reference_match|reference_match|3prime_fragment|3prime_fragment|alternative_3end|alternative_5end|alternative_5end|combination_of_known_junctions|alternative_3end5end|3prime_fragment|reference_match\n",
      "EP400\t-0.280616191\t0.038109071\t1\t0\t0\tcombination_of_known_junctions\n",
      "TRRAP\t-0.138325654\t0.414363477\t6\t0\t0\tinternal_fragment|internal_fragment|internal_fragment|internal_fragment|internal_fragment|internal_fragment\n",
      "BRD8\t0.157954396\t0.35760197\t15\t1\t0\tcombination_of_known_junctions|combination_of_known_junctions|3prime_fragment|3prime_fragment|reference_match|combination_of_known_junctions|reference_match|reference_match|combination_of_known_junctions|3prime_fragment|3prime_fragment|combination_of_known_junctions|3prime_fragment|combination_of_known_junctions|reference_match\n",
      "EPC1\t0.169092142\t0.139968173\t6\t1\t0\treference_match|reference_match|alternative_3end|5prime_fragment|5prime_fragment|5prime_fragment\n",
      "EPC2\t0.064374159\t0.701234555\t2\t0\t0\t3prime_fragment|3prime_fragment\n",
      "MBTD1\t-0.181013499\t0.07057304\t2\t0\t0\t5prime_fragment|at_least_one_novel_splicesite\n",
      "DMAP1\t-0.145453953\t0.279852815\t10\t0\t1\treference_match|combination_of_known_junctions|intron_retention|at_least_one_novel_splicesite|combination_of_known_splicesites|intron_retention|combination_of_known_junctions|alternative_5end|reference_match|3prime_fragment\n",
      "ING3\t0.024888729\t0.923187566\t7\t1\t1\treference_match|alternative_3end|alternative_3end|alternative_3end|alternative_3end|reference_match|alternative_3end\n",
      "VPS72\t-0.10319127\t0.650861628\t13\t2\t1\talternative_3end|combination_of_known_splicesites|at_least_one_novel_splicesite|3prime_fragment|3prime_fragment|alternative_3end5end|at_least_one_novel_splicesite|alternative_3end|at_least_one_novel_splicesite|combination_of_known_splicesites|3prime_fragment|alternative_3end|3prime_fragment\n",
      "RUVBL2\t0.111006825\t0.368152642\t16\t0\t0\tcombination_of_known_junctions|at_least_one_novel_splicesite|combination_of_known_junctions|combination_of_known_junctions|3prime_fragment|at_least_one_novel_splicesite|3prime_fragment|alternative_5end|3prime_fragment|at_least_one_novel_splicesite|3prime_fragment|alternative_5end|reference_match|3prime_fragment|3prime_fragment|3prime_fragment\n",
      "RUVBL1\t0.401894035\t0.000283264\t14\t0\t0\tat_least_one_novel_splicesite|combination_of_known_splicesites|3prime_fragment|reference_match|combination_of_known_splicesites|3prime_fragment|alternative_3end5end|3prime_fragment|alternative_3end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment\n",
      "ACTL6A\t0.131775332\t0.238432469\t14\t0\t0\talternative_3end5end|3prime_fragment|alternative_5end|reference_match|alternative_3end|reference_match|alternative_3end|alternative_3end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|alternative_3end|3prime_fragment\n",
      "ACTB\t0.158245079\t0.061976246\t30\t1\t0\talternative_3end|alternative_5end|reference_match|reference_match|intron_retention|intron_retention|alternative_5end|alternative_5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|alternative_3end|3prime_fragment|3prime_fragment|3prime_fragment|intron_retention|at_least_one_novel_splicesite|alternative_5end|3prime_fragment|3prime_fragment|3prime_fragment|alternative_5end|alternative_5end|alternative_3end|3prime_fragment|intron_retention|alternative_3end|3prime_fragment|alternative_5end\n",
      "MORF4L1\t0.238586326\t0.082911859\t33\t0\t0\tcombination_of_known_junctions|alternative_3end|combination_of_known_splicesites|at_least_one_novel_splicesite|alternative_3end5end|alternative_3end5end|alternative_3end|3prime_fragment|at_least_one_novel_splicesite|alternative_3end5end|3prime_fragment|alternative_3end5end|alternative_3end5end|3prime_fragment|alternative_3end|5prime_fragment|alternative_3end|combination_of_known_junctions|3prime_fragment|3prime_fragment|3prime_fragment|alternative_3end5end|alternative_3end5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|combination_of_known_junctions|3prime_fragment|3prime_fragment|alternative_3end5end|alternative_3end5end\n",
      "MORF4L2\t-0.59104297\t1.92E-11\t18\t1\t4\treference_match|alternative_3end5end|alternative_3end|alternative_3end|combination_of_known_splicesites|alternative_3end5end|alternative_3end5end|alternative_5end|alternative_3end5end|alternative_5end|3prime_fragment|3prime_fragment|alternative_5end|combination_of_known_splicesites|alternative_3end|combination_of_known_junctions|reference_match|alternative_3end\n",
      "YEATS4\t-0.044416731\t0.783745789\t8\t0\t2\talternative_3end|alternative_3end|alternative_3end|alternative_3end5end|3prime_fragment|at_least_one_novel_splicesite|3prime_fragment|alternative_3end5end\n",
      "MRGBP\t0.172190429\t0.417494383\t3\t0\t0\talternative_3end|3prime_fragment|alternative_3end5end\n",
      "MEAF6\t0.113594461\t0.368049763\t30\t2\t7\talternative_3end|alternative_3end|alternative_3end|alternative_3end|alternative_3end|alternative_3end5end|alternative_3end|alternative_3end|reference_match|3prime_fragment|alternative_3end|reference_match|alternative_3end5end|alternative_3end|3prime_fragment|3prime_fragment|alternative_3end|3prime_fragment|alternative_3end|alternative_3end5end|3prime_fragment|alternative_3end5end|alternative_3end|5prime_fragment|3prime_fragment|3prime_fragment|alternative_3end5end|alternative_3end5end|3prime_fragment|alternative_3end\n",
      "MYCL\t0.378667105\t3.24E-09\t15\t0\t0\treference_match|reference_match|alternative_5end|reference_match|alternative_5end|alternative_5end|alternative_5end|alternative_5end|at_least_one_novel_splicesite|alternative_5end|alternative_5end|alternative_5end|alternative_3end5end|alternative_3end|alternative_3end5end\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Define table structure\n",
    "table_data = []\n",
    "\n",
    "# Print header\n",
    "header = [\"Gene\", \"log2FoldChange\", \"padj\", \"num_isoforms\", \"significant_up\", \"significant_down\", \"categories\"]\n",
    "\n",
    "for cur_gene in genes:\n",
    "    illumina_fold = illumina_lookup[cur_gene][2]\n",
    "    illumina_p = illumina_lookup[cur_gene][5]\n",
    "\n",
    "    num_isoforms = len(pacbio_lookup[cur_gene])\n",
    "    categories = []\n",
    "    num_up = 0\n",
    "    num_down = 0\n",
    "\n",
    "    for isoform_results in pacbio_lookup[cur_gene]:\n",
    "        if float(isoform_results[5]) <= 0.05: # noramlly 5 for most tables\n",
    "            if float(isoform_results[2]) > 0:\n",
    "                num_up += 1\n",
    "            else:\n",
    "                num_down += 1\n",
    "        isoform = isoform_results[0]\n",
    "        classification = isoform_classification[isoform]\n",
    "        categories.append(classification)\n",
    "\n",
    "    # Append to table data\n",
    "    table_data.append([\n",
    "        cur_gene,\n",
    "        illumina_fold,  # Ensure numbers are formatted properly\n",
    "        illumina_p,\n",
    "        num_isoforms,\n",
    "        num_up,\n",
    "        num_down,\n",
    "        \"|\".join(categories)\n",
    "    ])\n",
    "\n",
    "# Create a DataFrame\n",
    "df = pd.DataFrame(table_data, columns=header)\n",
    "\n",
    "# # Format numeric columns for better readability\n",
    "# df[\"log2FoldChange\"] = df[\"log2FoldChange\"].map(lambda x: f\"{x:.3f}\")\n",
    "# df[\"padj\"] = df[\"padj\"].map(lambda x: f\"{x:.3e}\")  # Scientific notation for small p-values\n",
    "\n",
    "# Save to CSV (for sharing)\n",
    "df.to_csv(\"tip60_table.csv\", index=False)\n",
    "\n",
    "# Print a nicely formatted table manually\n",
    "print(\"\\t\".join(header))  # Print header row\n",
    "for row in table_data:\n",
    "    print(\"\\t\".join(map(str, row)))  # Print each row with tab separation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0e08bacf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene\tlog2FoldChange\tpadj\tnum_isoforms\tsignificant_up\tsignificant_down\tcategories\tsummary_category\n",
      "KAT5\t-0.410763045\t0.000225465\t13\t0\t2\treference_match|reference_match|alternative_5end|alternative_5end|alternative_3end|reference_match|combination_of_known_junctions|alternative_3end5end|alternative_5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment\tCanonical isoform|Truncated transcript|Other|Alternative splicing\n",
      "EP400\t-0.280616191\t0.038109071\t1\t0\t0\tcombination_of_known_junctions\tOther\n",
      "TRRAP\t-0.138325654\t0.414363477\t4\t0\t0\tinternal_fragment|internal_fragment|internal_fragment|internal_fragment\tOther\n",
      "BRD8\t0.157954396\t0.35760197\t15\t0\t0\treference_match|combination_of_known_junctions|combination_of_known_junctions|reference_match|combination_of_known_junctions|reference_match|combination_of_known_junctions|combination_of_known_junctions|combination_of_known_junctions|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|reference_match|3prime_fragment\tCanonical isoform|Truncated transcript|Other\n",
      "EPC1\t0.169092142\t0.139968173\t6\t1\t0\talternative_3end|5prime_fragment|5prime_fragment|5prime_fragment|reference_match|reference_match\tCanonical isoform|Truncated transcript|Alternative splicing\n",
      "EPC2\t0.064374159\t0.701234555\t4\t0\t0\treference_match|3prime_fragment|3prime_fragment|3prime_fragment\tCanonical isoform|Truncated transcript\n",
      "MBTD1\t-0.181013499\t0.07057304\t2\t0\t0\t5prime_fragment|at_least_one_novel_splicesite\tTruncated transcript|Novel isoform\n",
      "DMAP1\t-0.145453953\t0.279852815\t13\t0\t2\treference_match|combination_of_known_junctions|at_least_one_novel_splicesite|reference_match|combination_of_known_splicesites|combination_of_known_junctions|intron_retention|intron_retention|at_least_one_novel_splicesite|combination_of_known_junctions|alternative_5end|3prime_fragment|alternative_5end\tTruncated transcript|Alternative splicing|Other|Novel isoform|Canonical isoform\n",
      "ING3\t0.024888729\t0.923187566\t8\t1\t1\talternative_3end|alternative_3end|reference_match|reference_match|alternative_3end|alternative_3end|alternative_3end|alternative_3end\tCanonical isoform|Alternative splicing\n",
      "VPS72\t-0.10319127\t0.650861628\t13\t2\t1\talternative_3end|alternative_3end|at_least_one_novel_splicesite|combination_of_known_splicesites|at_least_one_novel_splicesite|at_least_one_novel_splicesite|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|combination_of_known_splicesites|alternative_3end5end|alternative_3end\tTruncated transcript|Other|Alternative splicing|Novel isoform\n",
      "RUVBL2\t0.111006825\t0.368152642\t15\t0\t0\tcombination_of_known_junctions|alternative_5end|combination_of_known_junctions|alternative_5end|reference_match|combination_of_known_junctions|at_least_one_novel_splicesite|at_least_one_novel_splicesite|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment\tTruncated transcript|Other|Alternative splicing|Novel isoform|Canonical isoform\n",
      "RUVBL1\t0.401894035\t0.000283264\t14\t0\t0\treference_match|alternative_3end|alternative_3end5end|at_least_one_novel_splicesite|combination_of_known_splicesites|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|combination_of_known_splicesites|3prime_fragment\tTruncated transcript|Other|Alternative splicing|Novel isoform|Canonical isoform\n",
      "ACTL6A\t0.131775332\t0.238432469\t13\t0\t0\treference_match|alternative_3end|reference_match|alternative_3end|alternative_3end|alternative_5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|alternative_3end5end\tCanonical isoform|Truncated transcript|Other|Alternative splicing\n",
      "ACTB\t0.158245079\t0.061976246\t33\t1\t0\treference_match|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|alternative_5end|alternative_3end|intron_retention|alternative_3end|alternative_5end|intron_retention|alternative_5end|alternative_5end|alternative_5end|reference_match|intron_retention|alternative_5end|alternative_5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|intron_retention|3prime_fragment|3prime_fragment|at_least_one_novel_splicesite|3prime_fragment|3prime_fragment|intron_retention|intron_retention|alternative_3end|alternative_3end|alternative_3end\tCanonical isoform|Truncated transcript|Alternative splicing|Novel isoform\n",
      "MORF4L1\t0.238586326\t0.082911859\t33\t0\t0\talternative_3end5end|alternative_3end5end|alternative_3end|alternative_3end|alternative_3end5end|alternative_3end5end|combination_of_known_junctions|5prime_fragment|alternative_3end|alternative_3end5end|alternative_3end5end|alternative_3end5end|alternative_3end5end|at_least_one_novel_splicesite|at_least_one_novel_splicesite|combination_of_known_junctions|combination_of_known_splicesites|combination_of_known_splicesites|combination_of_known_junctions|alternative_3end|alternative_3end5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment\tTruncated transcript|Other|Alternative splicing|Novel isoform\n",
      "MORF4L2\t-0.59104297\t1.92E-11\t18\t2\t4\talternative_3end5end|combination_of_known_splicesites|alternative_5end|alternative_3end|alternative_3end5end|alternative_3end|reference_match|alternative_3end|combination_of_known_junctions|alternative_5end|combination_of_known_splicesites|alternative_3end|alternative_5end|alternative_3end5end|reference_match|3prime_fragment|alternative_3end5end|3prime_fragment\tTruncated transcript|Alternative splicing|Other|Novel isoform|Canonical isoform\n",
      "YEATS4\t-0.044416731\t0.783745789\t7\t1\t2\talternative_3end|alternative_3end|alternative_3end|at_least_one_novel_splicesite|alternative_3end5end|3prime_fragment|3prime_fragment\tTruncated transcript|Other|Alternative splicing|Novel isoform\n",
      "MRGBP\t0.172190429\t0.417494383\t3\t0\t0\talternative_3end|alternative_3end5end|3prime_fragment\tTruncated transcript|Other|Alternative splicing\n",
      "MEAF6\t0.113594461\t0.368049763\t31\t2\t7\talternative_3end|alternative_3end|reference_match|alternative_3end|alternative_3end|alternative_3end|alternative_3end|alternative_3end|alternative_3end|alternative_3end5end|alternative_3end5end|alternative_3end5end|3prime_fragment|3prime_fragment|3prime_fragment|alternative_3end|alternative_3end5end|alternative_3end|alternative_3end|reference_match|alternative_3end5end|alternative_3end5end|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|3prime_fragment|alternative_3end|alternative_3end|alternative_3end|5prime_fragment\tCanonical isoform|Truncated transcript|Other|Alternative splicing\n",
      "H2AZ1\t0.040048488\t0.856458689\t10\t0\t0\talternative_5end|alternative_5end|3prime_fragment|3prime_fragment|intron_retention|intron_retention|alternative_5end|at_least_one_novel_splicesite|alternative_5end|alternative_5end\tTruncated transcript|Alternative splicing|Novel isoform\n",
      "MYCL\t0.378667105\t3.24E-09\t15\t0\t0\treference_match|alternative_5end|alternative_5end|reference_match|alternative_5end|alternative_5end|alternative_5end|alternative_5end|at_least_one_novel_splicesite|alternative_5end|alternative_3end5end|reference_match|alternative_5end|alternative_3end|alternative_3end5end\tCanonical isoform|Other|Alternative splicing|Novel isoform\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from collections import Counter\n",
    "\n",
    "# Define table structure\n",
    "table_data = []\n",
    "\n",
    "# Define broad isoform categories\n",
    "category_mapping = {\n",
    "    \"reference_match\": \"Canonical isoform\",\n",
    "    \"alternative_3end\": \"Alternative splicing\",\n",
    "    \"alternative_5end\": \"Alternative splicing\",\n",
    "    \"intron_retention\": \"Alternative splicing\",\n",
    "    \"at_least_one_novel_splicesite\": \"Novel isoform\",\n",
    "    \"combination_of_known_splicesites\": \"Novel isoform\",\n",
    "    \"3prime_fragment\": \"Truncated transcript\",\n",
    "    \"5prime_fragment\": \"Truncated transcript\"\n",
    "}\n",
    "\n",
    "# Function to determine the dominant isoform category for a gene\n",
    "def summarize_categories(categories):\n",
    "    mapped_categories = [category_mapping.get(cat, \"Other\") for cat in categories]\n",
    "    \n",
    "    unique_cats = set(mapped_categories)\n",
    "    \n",
    "    return '|'.join(unique_cats)\n",
    "#     most_common = Counter(mapped_categories).most_common(1)\n",
    "#     return most_common[0][0] if most_common else \"Unknown\"\n",
    "\n",
    "# Print header\n",
    "header = [\"Gene\", \"log2FoldChange\", \"padj\", \"num_isoforms\", \"significant_up\", \"significant_down\", \"categories\", \"summary_category\"]\n",
    "\n",
    "for cur_gene in genes:\n",
    "    illumina_fold = illumina_lookup[cur_gene][2]\n",
    "    illumina_p = illumina_lookup[cur_gene][5]\n",
    "\n",
    "    num_isoforms = len(pacbio_lookup[cur_gene])\n",
    "    categories = []\n",
    "    num_up = 0\n",
    "    num_down = 0\n",
    "\n",
    "    for isoform_results in pacbio_lookup[cur_gene]:\n",
    "        if float(isoform_results[5]) <= 0.05:  # Adjust significance threshold if needed\n",
    "            if float(isoform_results[2]) > 0:\n",
    "                num_up += 1\n",
    "            else:\n",
    "                num_down += 1\n",
    "        isoform = isoform_results[0].replace('\"', '')\n",
    "        classification = isoform_classification[isoform]\n",
    "        categories.append(classification)\n",
    "\n",
    "    # Determine dominant isoform category\n",
    "    summary_category = summarize_categories(categories)\n",
    "\n",
    "    # Append to table data\n",
    "    table_data.append([\n",
    "        cur_gene,\n",
    "        illumina_fold,  # Ensure numbers are formatted properly\n",
    "        illumina_p,\n",
    "        num_isoforms,\n",
    "        num_up,\n",
    "        num_down,\n",
    "        \"|\".join(categories),\n",
    "        summary_category\n",
    "    ])\n",
    "\n",
    "# Create a DataFrame\n",
    "df = pd.DataFrame(table_data, columns=header)\n",
    "\n",
    "# Save to CSV\n",
    "df.to_csv(\"tip60_table_summarized.csv\", index=False)\n",
    "\n",
    "# Print a nicely formatted table manually\n",
    "print(\"\\t\".join(header))  # Print header row\n",
    "for row in table_data:\n",
    "    print(\"\\t\".join(map(str, row)))  # Print each row with tab separation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fa28a84e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>padj</th>\n",
       "      <th>num_isoforms</th>\n",
       "      <th>significant_up</th>\n",
       "      <th>significant_down</th>\n",
       "      <th>categories</th>\n",
       "      <th>summary_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KAT5*</td>\n",
       "      <td>-0.410763045</td>\n",
       "      <td>0.000225465</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>reference_match|reference_match|alternative_5e...</td>\n",
       "      <td>Canonical isoform|Truncated transcript|Other|A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EP400*</td>\n",
       "      <td>-0.280616191</td>\n",
       "      <td>0.038109071</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>combination_of_known_junctions</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TRRAP</td>\n",
       "      <td>-0.138325654</td>\n",
       "      <td>0.414363477</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>internal_fragment|internal_fragment|internal_f...</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRD8</td>\n",
       "      <td>0.157954396</td>\n",
       "      <td>0.35760197</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>reference_match|combination_of_known_junctions...</td>\n",
       "      <td>Canonical isoform|Truncated transcript|Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EPC1</td>\n",
       "      <td>0.169092142</td>\n",
       "      <td>0.139968173</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>alternative_3end|5prime_fragment|5prime_fragme...</td>\n",
       "      <td>Canonical isoform|Truncated transcript|Alterna...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EPC2</td>\n",
       "      <td>0.064374159</td>\n",
       "      <td>0.701234555</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>reference_match|3prime_fragment|3prime_fragmen...</td>\n",
       "      <td>Canonical isoform|Truncated transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MBTD1</td>\n",
       "      <td>-0.181013499</td>\n",
       "      <td>0.07057304</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5prime_fragment|at_least_one_novel_splicesite</td>\n",
       "      <td>Truncated transcript|Novel isoform</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DMAP1</td>\n",
       "      <td>-0.145453953</td>\n",
       "      <td>0.279852815</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>reference_match|combination_of_known_junctions...</td>\n",
       "      <td>Truncated transcript|Alternative splicing|Othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ING3</td>\n",
       "      <td>0.024888729</td>\n",
       "      <td>0.923187566</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>alternative_3end|alternative_3end|reference_ma...</td>\n",
       "      <td>Canonical isoform|Alternative splicing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>VPS72</td>\n",
       "      <td>-0.10319127</td>\n",
       "      <td>0.650861628</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>alternative_3end|alternative_3end|at_least_one...</td>\n",
       "      <td>Truncated transcript|Other|Alternative splicin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>RUVBL2</td>\n",
       "      <td>0.111006825</td>\n",
       "      <td>0.368152642</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>combination_of_known_junctions|alternative_5en...</td>\n",
       "      <td>Truncated transcript|Other|Alternative splicin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>RUVBL1*</td>\n",
       "      <td>0.401894035</td>\n",
       "      <td>0.000283264</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>reference_match|alternative_3end|alternative_3...</td>\n",
       "      <td>Truncated transcript|Other|Alternative splicin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ACTL6A</td>\n",
       "      <td>0.131775332</td>\n",
       "      <td>0.238432469</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>reference_match|alternative_3end|reference_mat...</td>\n",
       "      <td>Canonical isoform|Truncated transcript|Other|A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>0.158245079</td>\n",
       "      <td>0.061976246</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>reference_match|3prime_fragment|3prime_fragmen...</td>\n",
       "      <td>Canonical isoform|Truncated transcript|Alterna...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MORF4L1</td>\n",
       "      <td>0.238586326</td>\n",
       "      <td>0.082911859</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>alternative_3end5end|alternative_3end5end|alte...</td>\n",
       "      <td>Truncated transcript|Other|Alternative splicin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MORF4L2*</td>\n",
       "      <td>-0.59104297</td>\n",
       "      <td>1.92E-11</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>alternative_3end5end|combination_of_known_spli...</td>\n",
       "      <td>Truncated transcript|Alternative splicing|Othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>YEATS4</td>\n",
       "      <td>-0.044416731</td>\n",
       "      <td>0.783745789</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>alternative_3end|alternative_3end|alternative_...</td>\n",
       "      <td>Truncated transcript|Other|Alternative splicin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MRGBP</td>\n",
       "      <td>0.172190429</td>\n",
       "      <td>0.417494383</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>alternative_3end|alternative_3end5end|3prime_f...</td>\n",
       "      <td>Truncated transcript|Other|Alternative splicing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MEAF6</td>\n",
       "      <td>0.113594461</td>\n",
       "      <td>0.368049763</td>\n",
       "      <td>31</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>alternative_3end|alternative_3end|reference_ma...</td>\n",
       "      <td>Canonical isoform|Truncated transcript|Other|A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>H2AZ1</td>\n",
       "      <td>0.040048488</td>\n",
       "      <td>0.856458689</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>alternative_5end|alternative_5end|3prime_fragm...</td>\n",
       "      <td>Truncated transcript|Alternative splicing|Nove...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MYCL*</td>\n",
       "      <td>0.378667105</td>\n",
       "      <td>3.24E-09</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>reference_match|alternative_5end|alternative_5...</td>\n",
       "      <td>Canonical isoform|Other|Alternative splicing|N...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene log2FoldChange         padj  num_isoforms  significant_up  \\\n",
       "0      KAT5*   -0.410763045  0.000225465            13               0   \n",
       "1     EP400*   -0.280616191  0.038109071             1               0   \n",
       "2      TRRAP   -0.138325654  0.414363477             4               0   \n",
       "3       BRD8    0.157954396   0.35760197            15               0   \n",
       "4       EPC1    0.169092142  0.139968173             6               1   \n",
       "5       EPC2    0.064374159  0.701234555             4               0   \n",
       "6      MBTD1   -0.181013499   0.07057304             2               0   \n",
       "7      DMAP1   -0.145453953  0.279852815            13               0   \n",
       "8       ING3    0.024888729  0.923187566             8               1   \n",
       "9      VPS72    -0.10319127  0.650861628            13               2   \n",
       "10    RUVBL2    0.111006825  0.368152642            15               0   \n",
       "11   RUVBL1*    0.401894035  0.000283264            14               0   \n",
       "12    ACTL6A    0.131775332  0.238432469            13               0   \n",
       "13      ACTB    0.158245079  0.061976246            33               1   \n",
       "14   MORF4L1    0.238586326  0.082911859            33               0   \n",
       "15  MORF4L2*    -0.59104297     1.92E-11            18               2   \n",
       "16    YEATS4   -0.044416731  0.783745789             7               1   \n",
       "17     MRGBP    0.172190429  0.417494383             3               0   \n",
       "18     MEAF6    0.113594461  0.368049763            31               2   \n",
       "19     H2AZ1    0.040048488  0.856458689            10               0   \n",
       "20     MYCL*    0.378667105     3.24E-09            15               0   \n",
       "\n",
       "    significant_down                                         categories  \\\n",
       "0                  2  reference_match|reference_match|alternative_5e...   \n",
       "1                  0                     combination_of_known_junctions   \n",
       "2                  0  internal_fragment|internal_fragment|internal_f...   \n",
       "3                  0  reference_match|combination_of_known_junctions...   \n",
       "4                  0  alternative_3end|5prime_fragment|5prime_fragme...   \n",
       "5                  0  reference_match|3prime_fragment|3prime_fragmen...   \n",
       "6                  0      5prime_fragment|at_least_one_novel_splicesite   \n",
       "7                  2  reference_match|combination_of_known_junctions...   \n",
       "8                  1  alternative_3end|alternative_3end|reference_ma...   \n",
       "9                  1  alternative_3end|alternative_3end|at_least_one...   \n",
       "10                 0  combination_of_known_junctions|alternative_5en...   \n",
       "11                 0  reference_match|alternative_3end|alternative_3...   \n",
       "12                 0  reference_match|alternative_3end|reference_mat...   \n",
       "13                 0  reference_match|3prime_fragment|3prime_fragmen...   \n",
       "14                 0  alternative_3end5end|alternative_3end5end|alte...   \n",
       "15                 4  alternative_3end5end|combination_of_known_spli...   \n",
       "16                 2  alternative_3end|alternative_3end|alternative_...   \n",
       "17                 0  alternative_3end|alternative_3end5end|3prime_f...   \n",
       "18                 7  alternative_3end|alternative_3end|reference_ma...   \n",
       "19                 0  alternative_5end|alternative_5end|3prime_fragm...   \n",
       "20                 0  reference_match|alternative_5end|alternative_5...   \n",
       "\n",
       "                                     summary_category  \n",
       "0   Canonical isoform|Truncated transcript|Other|A...  \n",
       "1                                               Other  \n",
       "2                                               Other  \n",
       "3        Canonical isoform|Truncated transcript|Other  \n",
       "4   Canonical isoform|Truncated transcript|Alterna...  \n",
       "5              Canonical isoform|Truncated transcript  \n",
       "6                  Truncated transcript|Novel isoform  \n",
       "7   Truncated transcript|Alternative splicing|Othe...  \n",
       "8              Canonical isoform|Alternative splicing  \n",
       "9   Truncated transcript|Other|Alternative splicin...  \n",
       "10  Truncated transcript|Other|Alternative splicin...  \n",
       "11  Truncated transcript|Other|Alternative splicin...  \n",
       "12  Canonical isoform|Truncated transcript|Other|A...  \n",
       "13  Canonical isoform|Truncated transcript|Alterna...  \n",
       "14  Truncated transcript|Other|Alternative splicin...  \n",
       "15  Truncated transcript|Alternative splicing|Othe...  \n",
       "16  Truncated transcript|Other|Alternative splicin...  \n",
       "17    Truncated transcript|Other|Alternative splicing  \n",
       "18  Canonical isoform|Truncated transcript|Other|A...  \n",
       "19  Truncated transcript|Alternative splicing|Nove...  \n",
       "20  Canonical isoform|Other|Alternative splicing|N...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from collections import Counter\n",
    "\n",
    "# Define table structure\n",
    "table_data = []\n",
    "\n",
    "# Define broad isoform categories\n",
    "category_mapping = {\n",
    "    \"reference_match\": \"Canonical isoform\",\n",
    "    \"alternative_3end\": \"Alternative splicing\",\n",
    "    \"alternative_5end\": \"Alternative splicing\",\n",
    "    \"intron_retention\": \"Alternative splicing\",\n",
    "    \"at_least_one_novel_splicesite\": \"Novel isoform\",\n",
    "    \"combination_of_known_splicesites\": \"Novel isoform\",\n",
    "    \"3prime_fragment\": \"Truncated transcript\",\n",
    "    \"5prime_fragment\": \"Truncated transcript\"\n",
    "}\n",
    "\n",
    "# Function to determine the dominant isoform category for a gene\n",
    "def summarize_categories(categories):\n",
    "    mapped_categories = [category_mapping.get(cat, \"Other\") for cat in categories]\n",
    "    \n",
    "    unique_cats = set(mapped_categories)\n",
    "    \n",
    "    return '|'.join(unique_cats)\n",
    "\n",
    "# Print header\n",
    "header = [\"Gene\", \"log2FoldChange\", \"padj\", \"num_isoforms\", \"significant_up\", \"significant_down\", \"categories\", \"summary_category\"]\n",
    "\n",
    "for cur_gene in genes:\n",
    "    illumina_fold = illumina_lookup[cur_gene][2]\n",
    "    illumina_p = illumina_lookup[cur_gene][5]\n",
    "\n",
    "    num_isoforms = len(pacbio_lookup[cur_gene])\n",
    "    categories = []\n",
    "    num_up = 0\n",
    "    num_down = 0\n",
    "\n",
    "    for isoform_results in pacbio_lookup[cur_gene]:\n",
    "        if float(isoform_results[5]) <= 0.05:  # Adjust significance threshold if needed\n",
    "            if float(isoform_results[2]) > 0:\n",
    "                num_up += 1\n",
    "            else:\n",
    "                num_down += 1\n",
    "        isoform = isoform_results[0].replace('\"', '')\n",
    "        classification = isoform_classification[isoform]\n",
    "        categories.append(classification)\n",
    "\n",
    "    # Add a star to the gene name if padj < 0.05\n",
    "    gene_name = cur_gene\n",
    "    if float(illumina_p) < 0.05:\n",
    "        gene_name = f\"{cur_gene}*\"\n",
    "\n",
    "    # Determine dominant isoform category\n",
    "    summary_category = summarize_categories(categories)\n",
    "\n",
    "    # Append to table data\n",
    "    table_data.append([gene_name,\n",
    "        illumina_fold,  # Ensure numbers are formatted properly\n",
    "        illumina_p,\n",
    "        num_isoforms,\n",
    "        num_up,\n",
    "        num_down,\n",
    "        \"|\".join(categories),\n",
    "        summary_category\n",
    "    ])\n",
    "\n",
    "# Create a DataFrame\n",
    "df = pd.DataFrame(table_data, columns=header)\n",
    "\n",
    "# Save to CSV\n",
    "df.to_csv(\"tip60_table_summarized.csv\", index=False)\n",
    "\n",
    "# Display the table without the row numbers\n",
    "df\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
